File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
from collections import defaultdict

def solution(genres, plays):
    answer = []
    genre_total = defaultdict(int)
    genre_songs = defaultdict(list)
    
    for i, (g, p) in enumerate(zip(genres, plays)):
        genre_total[g] += p
        genre_songs[g].append((p, i))
    
    sorted_genres = sorted(genre_total.items(), key=lambda x: -x[1]) ## 내림차순
    
    for gen in sorted_genres:
        songs = genre_songs[gen[0]]
        songs.sort(key=lambda x: -x[0])
        
        if len(songs) >= 2:
            answer.append(songs[0][1])
            answer.append(songs[1][1])
        elif len(songs) == 1:
            answer.append(songs[0][1])
    
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        List<Integer> result = new ArrayList<>();
        
        // 장르, 총 재생 수
        Map<String, Integer> genPlay = new HashMap<>();
        // 장르, 노래 리스트(id, 재생 수)
        Map<String, List<int []>> genSong = new HashMap<>();
        
        for(int i=0; i<genres.length; ++i){
            genPlay.put(genres[i], genPlay.getOrDefault(genres[i], 0)+plays[i]);
            
            genSong.putIfAbsent(genres[i], new ArrayList<>());
            genSong.get(genres[i]).add(new int[]{i, plays[i]});
        }
        
        List<String> sortGenres = new ArrayList<>(genPlay.keySet());
        sortGenres.sort((a, b) -> genPlay.get(b) - genPlay.get(a)); // 내림차순
        
        for(String gen: sortGenres){
            List<int []> songs = new ArrayList<>(genSong.get(gen));
            songs.sort((a, b) -> b[1] - a[1]); // 내림차순
            
            if (songs.size() >= 2){
                result.add(songs.get(0)[0]);
                result.add(songs.get(1)[0]);
            }
            else if (songs.size() == 1){
                result.add(songs.get(0)[0]);
            }
        }
        
        
        
        int[] answer = new int[result.size()];
        
        for(int i=0; i<result.size(); ++i){
            answer[i] = result.get(i);
        }
        
        
        return answer;
    }
}